File: src/investraton/engine/concentration.py
"""Concentration & overlap analysis on top of look-through exposure.

Surfaces the things the plan called out: hidden single-name concentration (you own
a stock directly AND via your ETFs), ETF overlap (two funds holding the same
names), and region/sector concentration. Produces Findings — descriptive, with
why-it-matters context, never advice.
"""

from __future__ import annotations

from itertools import combinations

from ..models import Finding
from .lookthrough import LookThrough

# Tunable thresholds (kept here, explicit and transparent).
SINGLE_NAME_NOTE = 0.03      # effective weight that's worth a mention
SINGLE_NAME_WATCH = 0.05     # effective weight worth a stronger flag
HIDDEN_MIN = 0.02            # min combined weight to flag direct+ETF overlap of one name
OVERLAP_NOTE = 0.15          # ETF-vs-ETF shared fraction worth mentioning
COUNTRY_NOTE = 0.50          # top-country share worth a mention
SECTOR_NOTE = 0.25           # top-sector share worth a mention


def analyse_concentration(lt: LookThrough) -> list[Finding]:
    findings: list[Finding] = []
    findings += _single_name(lt)
    findings += _hidden(lt)
    findings += _overlap(lt)
    findings += _geography(lt)
    findings += _sector(lt)
    return findings


def _single_name(lt: LookThrough) -> list[Finding]:
    top = [e for e in lt.by_name if e.weight > 0][:5]
    if not top:
        return []
    leader = top[0]
    sev = "watch" if leader.weight >= SINGLE_NAME_WATCH else ("note" if leader.weight >= SINGLE_NAME_NOTE else "info")
    listing = ", ".join(f"{e.label} {e.weight*100:.1f}%" for e in top)
    return [Finding(
        kind="concentration",
        title=f"Top effective holdings: {leader.label} at {leader.weight*100:.1f}%",
        detail=f"After looking through your ETFs, your largest single-name exposures are: {listing}. "
               f"This is your true company-level concentration, not what the ticker list suggests.",
        severity=sev,
    )]


def _hidden(lt: LookThrough) -> list[Finding]:
    hidden = [e for e in lt.by_name if e.direct > 0 and e.via_etf > 0 and e.weight >= HIDDEN_MIN]
    hidden.sort(key=lambda e: e.weight, reverse=True)
    out = []
    for e in hidden[:4]:
        out.append(Finding(
            kind="concentration",
            title=f"Hidden concentration in {e.label}: {e.weight*100:.1f}% effective",
            detail=f"You hold {e.label} both directly ({e.direct*100:.1f}%) and inside your ETFs "
                   f"({e.via_etf*100:.1f}%). Your real exposure is {e.weight*100:.1f}% — more than the "
                   f"direct position alone implies.",
            severity="watch" if e.weight >= SINGLE_NAME_WATCH else "note",
        ))
    return out


def _overlap(lt: LookThrough) -> list[Finding]:
    out = []
    etfs = lt.etf_constituents
    for a, b in combinations(sorted(etfs), 2):
        wa, wb = etfs[a], etfs[b]
        common = set(wa) & set(wb)
        if not common:
            continue
        shared = sum(min(wa[t], wb[t]) for t in common)
        if shared >= OVERLAP_NOTE:
            out.append(Finding(
                kind="overlap",
                title=f"{a} and {b} overlap ~{shared*100:.0f}%",
                detail=f"{a} and {b} share roughly {shared*100:.0f}% of their holdings ({len(common)} common "
                       f"names). Holding both gives you less diversification than two funds suggest — you're "
                       f"doubling down on the shared names.",
                severity="note",
            ))
    return out


def _geography(lt: LookThrough) -> list[Finding]:
    regions = [e for e in lt.by_region if e.key not in ("Unmapped ETF",)]
    decoded = sum(e.weight for e in regions)
    if decoded <= 0:
        return []
    top = max(regions, key=lambda e: e.weight)
    share = top.weight / decoded
    reg3 = ", ".join(f"{e.label} {e.weight/decoded*100:.0f}%"
                     for e in sorted(regions, key=lambda e: e.weight, reverse=True)[:3])
    countries = [e for e in lt.by_country if e.key not in ("Unmapped ETF",)]
    cty3 = ", ".join(f"{e.label} {e.weight/sum(c.weight for c in countries)*100:.0f}%"
                     for e in sorted(countries, key=lambda e: e.weight, reverse=True)[:3]) if countries else ""
    sev = "note" if share >= COUNTRY_NOTE else "info"
    return [Finding(
        kind="region",
        title=f"Regional tilt: {top.label} {share*100:.0f}% of decoded exposure",
        detail=f"By region: {reg3}." + (f" Top countries: {cty3}." if cty3 else "")
               + " A regional event (e.g. across Europe or Asia) hits every name in that bloc, "
                 "so this is your real geographic concentration.",
        severity=sev,
    )]


def _sector(lt: LookThrough) -> list[Finding]:
    sec = [e for e in lt.by_sector if e.key not in ("Unmapped ETF",)]
    decoded = sum(e.weight for e in sec)
    if decoded <= 0:
        return []
    top = max(sec, key=lambda e: e.weight)
    share = top.weight / decoded
    top3 = ", ".join(f"{e.label} {e.weight/decoded*100:.0f}%" for e in sorted(sec, key=lambda e: e.weight, reverse=True)[:3])
    sev = "note" if share >= SECTOR_NOTE else "info"
    return [Finding(
        kind="sector",
        title=f"Sector tilt: {top.label} {share*100:.0f}% of decoded exposure",
        detail=f"Your largest sector exposures are {top3}. Worth knowing when a sector-specific event hits.",
        severity=sev,
    )]

File: engine/src/concentration.ts
/** Port of engine/concentration.py — concentration & overlap analysis.
 *
 * Surfaces hidden single-name concentration (owned directly AND via ETFs), ETF
 * overlap, and region/sector tilt. Descriptive findings, never advice.
 */

import type { Exposure, Finding } from "./models.ts";
import type { LookThrough } from "./lookthrough.ts";
import { pyFormat } from "./round.ts";

// Tunable thresholds — explicit and transparent.
const SINGLE_NAME_NOTE = 0.03;
const SINGLE_NAME_WATCH = 0.05;
const HIDDEN_MIN = 0.02;
const OVERLAP_NOTE = 0.15;
const COUNTRY_NOTE = 0.50;
const SECTOR_NOTE = 0.25;

const f1 = (x: number) => pyFormat(x, 1);
const f0 = (x: number) => pyFormat(x, 0);

export function analyseConcentration(lt: LookThrough): Finding[] {
  return [...singleName(lt), ...hidden(lt), ...overlap(lt), ...geography(lt), ...sector(lt)];
}

function singleName(lt: LookThrough): Finding[] {
  const top = lt.by_name.filter((e) => e.weight > 0).slice(0, 5);
  if (top.length === 0) return [];
  const leader = top[0];
  const sev = leader.weight >= SINGLE_NAME_WATCH ? "watch"
    : leader.weight >= SINGLE_NAME_NOTE ? "note" : "info";
  const listing = top.map((e) => `${e.label} ${f1(e.weight * 100)}%`).join(", ");
  return [{
    kind: "concentration",
    title: `Top effective holdings: ${leader.label} at ${f1(leader.weight * 100)}%`,
    detail: `After looking through your ETFs, your largest single-name exposures are: ${listing}. ` +
      `This is your true company-level concentration, not what the ticker list suggests.`,
    severity: sev,
  }];
}

function hidden(lt: LookThrough): Finding[] {
  const items = lt.by_name
    .filter((e) => e.direct > 0 && e.via_etf > 0 && e.weight >= HIDDEN_MIN)
    .sort((a, b) => b.weight - a.weight)
    .slice(0, 4);
  return items.map((e) => ({
    kind: "concentration",
    title: `Hidden concentration in ${e.label}: ${f1(e.weight * 100)}% effective`,
    detail: `You hold ${e.label} both directly (${f1(e.direct * 100)}%) and inside your ETFs ` +
      `(${f1(e.via_etf * 100)}%). Your real exposure is ${f1(e.weight * 100)}% — more than the ` +
      `direct position alone implies.`,
    severity: e.weight >= SINGLE_NAME_WATCH ? "watch" : "note",
  }));
}

function overlap(lt: LookThrough): Finding[] {
  const out: Finding[] = [];
  const etfs = lt.etf_constituents;
  const names = Object.keys(etfs).sort();
  for (let i = 0; i < names.length; i++) {
    for (let j = i + 1; j < names.length; j++) {
      const a = names[i], b = names[j];
      const wa = etfs[a], wb = etfs[b];
      const common = Object.keys(wa).filter((t) => t in wb);
      if (common.length === 0) continue;
      const shared = common.reduce((acc, t) => acc + Math.min(wa[t], wb[t]), 0);
      if (shared >= OVERLAP_NOTE) {
        out.push({
          kind: "overlap",
          title: `${a} and ${b} overlap ~${f0(shared * 100)}%`,
          detail: `${a} and ${b} share roughly ${f0(shared * 100)}% of their holdings ` +
            `(${common.length} common names). Holding both gives you less diversification than ` +
            `two funds suggest — you're doubling down on the shared names.`,
          severity: "note",
        });
      }
    }
  }
  return out;
}

function geography(lt: LookThrough): Finding[] {
  const regions = lt.by_region.filter((e) => e.key !== "Unmapped ETF");
  const decoded = regions.reduce((a, e) => a + e.weight, 0);
  if (decoded <= 0) return [];
  const top = regions.reduce((m, e) => (e.weight > m.weight ? e : m), regions[0]);
  const share = top.weight / decoded;
  const reg3 = [...regions].sort((a, b) => b.weight - a.weight).slice(0, 3)
    .map((e) => `${e.label} ${f0(e.weight / decoded * 100)}%`).join(", ");
  const countries = lt.by_country.filter((e) => e.key !== "Unmapped ETF");
  const cTotal = countries.reduce((a, e) => a + e.weight, 0);
  const cty3 = countries.length
    ? [...countries].sort((a, b) => b.weight - a.weight).slice(0, 3)
        .map((e) => `${e.label} ${f0(e.weight / cTotal * 100)}%`).join(", ")
    : "";
  return [{
    kind: "region",
    title: `Regional tilt: ${top.label} ${f0(share * 100)}% of decoded exposure`,
    detail: `By region: ${reg3}.` + (cty3 ? ` Top countries: ${cty3}.` : "") +
      ` A regional event (e.g. across Europe or Asia) hits every name in that bloc, ` +
      `so this is your real geographic concentration.`,
    severity: share >= COUNTRY_NOTE ? "note" : "info",
  }];
}

function sector(lt: LookThrough): Finding[] {
  const sec = lt.by_sector.filter((e) => e.key !== "Unmapped ETF");
  const decoded = sec.reduce((a, e) => a + e.weight, 0);
  if (decoded <= 0) return [];
  const top = sec.reduce((m, e) => (e.weight > m.weight ? e : m), sec[0]);
  const share = top.weight / decoded;
  const top3 = [...sec].sort((a, b) => b.weight - a.weight).slice(0, 3)
    .map((e) => `${e.label} ${f0(e.weight / decoded * 100)}%`).join(", ");
  return [{
    kind: "sector",
    title: `Sector tilt: ${top.label} ${f0(share * 100)}% of decoded exposure`,
    detail: `Your largest sector exposures are ${top3}. Worth knowing when a sector-specific event hits.`,
    severity: share >= SECTOR_NOTE ? "note" : "info",
  }];
}

export type { Exposure };
